experimental/FTheoryTools/src/AbstractFTheoryModels/attributes.jl

```diff
@@ -1195,33 +1186,222 @@ function chern_class_c2(m::AbstractFTheoryModel; check::Bool = true)
+toric ambient space $X_\Sigma$, such that its restriction to $Y_n$ is the 4th Chern class
+$c_4$ of the tangent bundle of $Y_n$. If those assumptions are satisfied, this method returns
+this very cohomology class $h$, otherwise it raises and error.
+
+As of right now, this method is computationally quite expensive for involved toric varieties,
+such as in the example below. Therefore, think carefully if you truly want to compute this quantity.
+
+The computation of the cohomology ring verifies if the toric variety is simplicial and
+complete. The check for it to be complete can be very time consuming. This can be switched
+off by setting the optional argument `check` to the value `false`, as in the example below.
+
+```jldoctest; setup = :(Oscar.LazyArtifacts.ensure_artifact_installed("QSMDB", Oscar.LazyArtifacts.find_artifacts_toml(Oscar.oscardir)))
+julia> qsm_model = literature_model(arxiv_id = "1903.00009", model_parameters = Dict("k" => 4))
+Hypersurface model over a concrete base
+
+julia> h = chern_class_c4(qsm_model; check = false);
+
+julia> is_trivial(h)
+false
+```
+"""
+function chern_class_c4(m::AbstractFTheoryModel; check::Bool = true)
+  @req (m isa WeierstrassModel || m isa GlobalTateModel || m isa HypersurfaceModel) "Fourth Chern class of F-theory model supported for Weierstrass, global Tate and hypersurface models only"
+  @req base_space(m) isa NormalToricVariety "Fourth Chern class of F-theory model currently supported only for toric base"
+  @req ambient_space(m) isa NormalToricVariety "Fourth Chern class of F-theory model currently supported only for toric ambient space"
+
+  # Check if the answer is known
+  if has_attribute(m, :chern_class_c4)
+    return get_attribute(m, :chern_class_c4)
+  end
+
+  # Compute elements of the cohomology ring corresponding to the torus invariant prime divisors
+  q = Oscar._cohomology_ring_with_quotient_map(ambient_space(m); check = false)
+  c_ds = [g.f for g in gens(q[1])]
+  R = parent(c_ds[1])
+  projection_map = q[2]
+  len_c_ds = length(c_ds)
+
+  # Compute the cohomology class corresponding to the hypersurface equation
+  cy = cohomology_class(toric_divisor_class(ambient_space(m), degree(hypersurface_equation(m))))
+
+  # Identify index combination that may be discarded by virtue of the Stanley-Reisner ideal
+  mnf = Oscar._minimal_nonfaces(ambient_space(m))
+  indices = [findall(x -> x == 1, Vector{Int}(mnf[i, :])) for i in 1:Polymake.nrows(mnf)]
+
+  # Compute c4 of the toric ambient space
+  c4_t_ambient = zero(R)
+  for i in 1:len_c_ds - 3
+    for j in i+1:len_c_ds - 2
+      [i,j] in indices && continue
+      for k in j+1:len_c_ds - 1
+        ([i,k] in indices || [j,k] in indices) && continue
+        [i,j,k] in indices && continue
+        for l in k+1:len_c_ds
+          ([i,j,l] in indices || [i,k,l] in indices || [j,k,l] in indices) && continue
+          [i,j,k,l] in indices && continue
```

Comment:
It seems to me you are looking at all subsets of `[i,j,k,l]` of size >= 2.

Maybe could merge your c2/c3/c4 methods by exploiting that? It seems you are simply iterating over subsets of `1:len_c_ds` of size 2/3/4 that do not occur in indices, and for which no subsets occur in indices?

And then for every tuple `t` you find, you do `cN_t_ambient += prod(i -> c_cs[i], t)` ?

Reply:
Yes, that is exactly the task at hand.

For the k-th Chern class of the tangent bundle of a toric space, we would look at the subsets of exactly k elements (and returning 0 if k exceeds the dimension of the space). So for this generalization, this approach would help.

Currently, this loop is the bottle neck in this computation. I am not (yet) sure what can be done to speed it up further.

For the example in the doctests here, it is fast enough (as of my latest check, the computation of c4 required about 5 seconds on my laptop). But I have my eyes on a toric spaces, for which the indices i,j,k,l each can take values between roughly 1 and 300. And then, this method effectively does not terminate any more.